File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/ExceptionMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.TerminalErrorCode
import com.stripe.stripeterminal.external.models.TerminalException
import mek.stripeterminal.api.PlatformError
import mek.stripeterminal.api.TerminalExceptionApi
import mek.stripeterminal.api.TerminalExceptionCodeApi

fun TerminalException.toPlatformError(): PlatformError {
    return toApi().toPlatformError()
}

fun TerminalExceptionApi.toPlatformError(): PlatformError {
    return PlatformError(code = "mek_stripe_terminal", message = null, details = serialize())
}

fun TerminalException.toApi(): TerminalExceptionApi {
    val code = errorCode.toApiCode()
    return TerminalExceptionApi(
        code = code ?: TerminalExceptionCodeApi.UNKNOWN,
        message = errorMessage,
        stackTrace = stackTraceToString(),
        paymentIntent = paymentIntent?.toApi(),
        apiError = apiError?.toString()
    )
}

private fun TerminalErrorCode.toApiCode(): TerminalExceptionCodeApi? {
    return when (this) {
        TerminalErrorCode.CANCEL_FAILED -> TerminalExceptionCodeApi.CANCEL_FAILED
        TerminalErrorCode.NOT_CONNECTED_TO_READER -> TerminalExceptionCodeApi.NOT_CONNECTED_TO_READER
        TerminalErrorCode.ALREADY_CONNECTED_TO_READER -> TerminalExceptionCodeApi.ALREADY_CONNECTED_TO_READER
        TerminalErrorCode.BLUETOOTH_PERMISSION_DENIED -> TerminalExceptionCodeApi.BLUETOOTH_PERMISSION_DENIED
        TerminalErrorCode.CONFIRM_INVALID_PAYMENT_INTENT -> TerminalExceptionCodeApi.CONFIRM_INVALID_PAYMENT_INTENT
        TerminalErrorCode.CONFIRM_INVALID_SETUP_INTENT -> TerminalExceptionCodeApi.CONFIRM_INVALID_SETUP_INTENT
        TerminalErrorCode.INVALID_CLIENT_SECRET -> TerminalExceptionCodeApi.INVALID_CLIENT_SECRET
        TerminalErrorCode.UNSUPPORTED_OPERATION -> TerminalExceptionCodeApi.UNSUPPORTED_OPERATION
        TerminalErrorCode.UNEXPECTED_OPERATION -> TerminalExceptionCodeApi.UNEXPECTED_OPERATION
        TerminalErrorCode.UNSUPPORTED_SDK -> TerminalExceptionCodeApi.UNSUPPORTED_SDK
        TerminalErrorCode.USB_PERMISSION_DENIED -> TerminalExceptionCodeApi.USB_PERMISSION_DENIED
        TerminalErrorCode.MISSING_PREREQUISITE -> null
        TerminalErrorCode.MISSING_REQUIRED_PARAMETER -> TerminalExceptionCodeApi.INVALID_PARAMETER
        TerminalErrorCode.INVALID_REQUIRED_PARAMETER -> TerminalExceptionCodeApi.INVALID_REQUIRED_PARAMETER
        TerminalErrorCode.INVALID_TIP_PARAMETER -> TerminalExceptionCodeApi.INVALID_TIP_PARAMETER
        TerminalErrorCode.TAP_TO_PAY_LIBRARY_NOT_INCLUDED -> null
        TerminalErrorCode.TAP_TO_PAY_UNSUPPORTED_DEVICE -> TerminalExceptionCodeApi.TAP_TO_PAY_UNSUPPORTED_DEVICE
        TerminalErrorCode.TAP_TO_PAY_UNSUPPORTED_ANDROID_VERSION -> TerminalExceptionCodeApi.TAP_TO_PAY_UNSUPPORTED_OPERATING_SYSTEM_VERSION
        TerminalErrorCode.TAP_TO_PAY_DEVICE_TAMPERED -> TerminalExceptionCodeApi.TAP_TO_PAY_DEVICE_TAMPERED
        TerminalErrorCode.TAP_TO_PAY_DEBUG_NOT_SUPPORTED -> TerminalExceptionCodeApi.TAP_TO_PAY_DEBUG_NOT_SUPPORTED
        TerminalErrorCode.OFFLINE_MODE_UNSUPPORTED_ANDROID_VERSION -> TerminalExceptionCodeApi.OFFLINE_MODE_UNSUPPORTED_OPERATING_SYSTEM_VERSION
        TerminalErrorCode.CANCELED -> TerminalExceptionCodeApi.CANCELED
        TerminalErrorCode.LOCATION_SERVICES_DISABLED -> TerminalExceptionCodeApi.LOCATION_SERVICES_DISABLED
        TerminalErrorCode.BLUETOOTH_SCAN_TIMED_OUT -> TerminalExceptionCodeApi.BLUETOOTH_SCAN_TIMED_OUT
        TerminalErrorCode.BLUETOOTH_LOW_ENERGY_UNSUPPORTED -> TerminalExceptionCodeApi.BLUETOOTH_LOW_ENERGY_UNSUPPORTED
        TerminalErrorCode.READER_SOFTWARE_UPDATE_FAILED_BATTERY_LOW -> TerminalExceptionCodeApi.READER_SOFTWARE_UPDATE_FAILED_BATTERY_LOW
        TerminalErrorCode.READER_SOFTWARE_UPDATE_FAILED_INTERRUPTED -> TerminalExceptionCodeApi.READER_SOFTWARE_UPDATE_FAILED_INTERRUPTED
        TerminalErrorCode.CARD_INSERT_NOT_READ -> TerminalExceptionCodeApi.CARD_INSERT_NOT_READ
        TerminalErrorCode.CARD_SWIPE_NOT_READ -> TerminalExceptionCodeApi.CARD_SWIPE_NOT_READ
        TerminalErrorCode.CARD_READ_TIMED_OUT -> TerminalExceptionCodeApi.CARD_READ_TIMED_OUT
        TerminalErrorCode.CARD_REMOVED -> TerminalExceptionCodeApi.CARD_REMOVED
        TerminalErrorCode.CUSTOMER_CONSENT_REQUIRED -> TerminalExceptionCodeApi.CUSTOMER_CONSENT_REQUIRED
        TerminalErrorCode.CARD_LEFT_IN_READER -> TerminalExceptionCodeApi.CARD_LEFT_IN_READER
        TerminalErrorCode.USB_DISCOVERY_TIMED_OUT -> TerminalExceptionCodeApi.USB_DISCOVERY_TIMED_OUT
        TerminalErrorCode.FEATURE_NOT_ENABLED_ON_ACCOUNT -> TerminalExceptionCodeApi.FEATURE_NOT_ENABLED_ON_ACCOUNT
        TerminalErrorCode.READER_BUSY -> TerminalExceptionCodeApi.READER_BUSY
        TerminalErrorCode.READER_COMMUNICATION_ERROR -> TerminalExceptionCodeApi.READER_COMMUNICATION_ERROR
        TerminalErrorCode.BLUETOOTH_ERROR -> TerminalExceptionCodeApi.BLUETOOTH_ERROR
        TerminalErrorCode.BLUETOOTH_DISCONNECTED -> TerminalExceptionCodeApi.BLUETOOTH_DISCONNECTED
        TerminalErrorCode.BLUETOOTH_RECONNECT_STARTED -> TerminalExceptionCodeApi.BLUETOOTH_RECONNECT_STARTED
        TerminalErrorCode.USB_DISCONNECTED -> TerminalExceptionCodeApi.USB_DISCONNECTED
        TerminalErrorCode.USB_RECONNECT_STARTED -> TerminalExceptionCodeApi.USB_RECONNECT_STARTED
        TerminalErrorCode.READER_CONNECTED_TO_ANOTHER_DEVICE -> TerminalExceptionCodeApi.READER_CONNECTED_TO_ANOTHER_DEVICE
        TerminalErrorCode.READER_SOFTWARE_UPDATE_FAILED -> TerminalExceptionCodeApi.READER_SOFTWARE_UPDATE_FAILED
        TerminalErrorCode.READER_SOFTWARE_UPDATE_FAILED_READER_ERROR -> TerminalExceptionCodeApi.READER_SOFTWARE_UPDATE_FAILED_READER_ERROR
        TerminalErrorCode.READER_SOFTWARE_UPDATE_FAILED_SERVER_ERROR -> TerminalExceptionCodeApi.READER_SOFTWARE_UPDATE_FAILED_SERVER_ERROR
        TerminalErrorCode.TAP_TO_PAY_NFC_DISABLED -> TerminalExceptionCodeApi.NFC_DISABLED
        TerminalErrorCode.UNSUPPORTED_READER_VERSION -> TerminalExceptionCodeApi.UNSUPPORTED_READER_VERSION
        TerminalErrorCode.UNEXPECTED_SDK_ERROR -> TerminalExceptionCodeApi.UNEXPECTED_SDK_ERROR
        TerminalErrorCode.DECLINED_BY_STRIPE_API -> TerminalExceptionCodeApi.DECLINED_BY_STRIPE_API
        TerminalErrorCode.DECLINED_BY_READER -> TerminalExceptionCodeApi.DECLINED_BY_READER
        TerminalErrorCode.REQUEST_TIMED_OUT -> TerminalExceptionCodeApi.REQUEST_TIMED_OUT
        TerminalErrorCode.STRIPE_API_CONNECTION_ERROR -> TerminalExceptionCodeApi.STRIPE_API_CONNECTION_ERROR
        TerminalErrorCode.STRIPE_API_ERROR -> TerminalExceptionCodeApi.STRIPE_API_ERROR
        TerminalErrorCode.STRIPE_API_RESPONSE_DECODING_ERROR -> TerminalExceptionCodeApi.STRIPE_API_RESPONSE_DECODING_ERROR
        TerminalErrorCode.CONNECTION_TOKEN_PROVIDER_ERROR -> TerminalExceptionCodeApi.CONNECTION_TOKEN_PROVIDER_ERROR
        TerminalErrorCode.SESSION_EXPIRED -> TerminalExceptionCodeApi.SESSION_EXPIRED
        TerminalErrorCode.ANDROID_API_LEVEL_ERROR -> TerminalExceptionCodeApi.UNSUPPORTED_MOBILE_DEVICE_CONFIGURATION
        TerminalErrorCode.AMOUNT_EXCEEDS_MAX_OFFLINE_AMOUNT -> TerminalExceptionCodeApi.AMOUNT_EXCEEDS_MAX_OFFLINE_AMOUNT
        TerminalErrorCode.OFFLINE_PAYMENTS_DATABASE_TOO_LARGE -> TerminalExceptionCodeApi.OFFLINE_PAYMENTS_DATABASE_TOO_LARGE
        TerminalErrorCode.READER_CONNECTION_NOT_AVAILABLE_OFFLINE -> TerminalExceptionCodeApi.READER_CONNECTION_NOT_AVAILABLE_OFFLINE
        TerminalErrorCode.LOCATION_CONNECTION_NOT_AVAILABLE_OFFLINE -> TerminalExceptionCodeApi.LOCATION_CONNECTION_NOT_AVAILABLE_OFFLINE
        TerminalErrorCode.NO_LAST_SEEN_ACCOUNT -> TerminalExceptionCodeApi.NO_LAST_SEEN_ACCOUNT
        TerminalErrorCode.INVALID_OFFLINE_CURRENCY -> TerminalExceptionCodeApi.INVALID_OFFLINE_CURRENCY
        TerminalErrorCode.CARD_SWIPE_NOT_AVAILABLE -> TerminalExceptionCodeApi.CARD_SWIPE_NOT_AVAILABLE
        TerminalErrorCode.INTERAC_NOT_SUPPORTED_OFFLINE -> TerminalExceptionCodeApi.INTERAC_NOT_SUPPORTED_OFFLINE
        TerminalErrorCode.ONLINE_PIN_NOT_SUPPORTED_OFFLINE -> TerminalExceptionCodeApi.ONLINE_PIN_NOT_SUPPORTED_OFFLINE
        TerminalErrorCode.MOBILE_WALLET_NOT_SUPPORTED_ON_SETUP_INTENTS -> TerminalExceptionCodeApi.MOBILE_WALLET_NOT_SUPPORTED_ON_SETUP_INTENTS
        TerminalErrorCode.OFFLINE_AND_CARD_EXPIRED -> TerminalExceptionCodeApi.OFFLINE_AND_CARD_EXPIRED
        TerminalErrorCode.OFFLINE_TRANSACTION_DECLINED -> TerminalExceptionCodeApi.OFFLINE_TRANSACTION_DECLINED
        TerminalErrorCode.OFFLINE_COLLECT_AND_CONFIRM_MISMATCH -> TerminalExceptionCodeApi.OFFLINE_COLLECT_AND_CONFIRM_MISMATCH
        TerminalErrorCode.OFFLINE_TESTMODE_PAYMENT_IN_LIVEMODE -> TerminalExceptionCodeApi.FORWARDING_TEST_MODE_PAYMENT_IN_LIVE_MODE
        TerminalErrorCode.OFFLINE_LIVEMODE_PAYMENT_IN_TESTMODE -> TerminalExceptionCodeApi.FORWARDING_LIVE_MODE_PAYMENT_IN_TEST_MODE
        TerminalErrorCode.OFFLINE_PAYMENT_INTENT_NOT_FOUND -> TerminalExceptionCodeApi.OFFLINE_PAYMENT_INTENT_NOT_FOUND
        TerminalErrorCode.MISSING_EMV_DATA -> TerminalExceptionCodeApi.MISSING_EMV_DATA
        TerminalErrorCode.CONNECTION_TOKEN_PROVIDER_ERROR_WHILE_FORWARDING -> TerminalExceptionCodeApi.CONNECTION_TOKEN_PROVIDER_ERROR_WHILE_FORWARDING
        TerminalErrorCode.ACCOUNT_ID_MISMATCH_WHILE_FORWARDING -> TerminalExceptionCodeApi.ACCOUNT_ID_MISMATCH_WHILE_FORWARDING
        TerminalErrorCode.FORCE_OFFLINE_WITH_FEATURE_DISABLED -> TerminalExceptionCodeApi.OFFLINE_BEHAVIOR_FORCE_OFFLINE_WITH_FEATURE_DISABLED
        TerminalErrorCode.NOT_CONNECTED_TO_INTERNET_AND_REQUIRE_ONLINE_SET -> TerminalExceptionCodeApi.NOT_CONNECTED_TO_INTERNET_AND_OFFLINE_BEHAVIOR_REQUIRE_ONLINE
        TerminalErrorCode.TEST_CARD_IN_LIVEMODE -> TerminalExceptionCodeApi.TEST_CARD_IN_LIVE_MODE
        TerminalErrorCode.COLLECT_INPUTS_APPLICATION_ERROR -> TerminalExceptionCodeApi.COLLECT_INPUTS_APPLICATION_ERROR
        TerminalErrorCode.COLLECT_INPUTS_TIMED_OUT -> TerminalExceptionCodeApi.COLLECT_INPUTS_TIMED_OUT
        TerminalErrorCode.COLLECT_INPUTS_INVALID_PARAMETER -> TerminalExceptionCodeApi.INVALID_PARAMETER
        TerminalErrorCode.COLLECT_INPUTS_UNSUPPORTED -> TerminalExceptionCodeApi.COLLECT_INPUTS_UNSUPPORTED
        TerminalErrorCode.READER_BATTERY_CRITICALLY_LOW -> TerminalExceptionCodeApi.READER_BATTERY_CRITICALLY_LOW
        TerminalErrorCode.READER_SETTINGS_ERROR -> TerminalExceptionCodeApi.READER_SETTINGS_ERROR
        TerminalErrorCode.READER_MISSING_ENCRYPTION_KEYS -> TerminalExceptionCodeApi.READER_MISSING_ENCRYPTION_KEYS
        TerminalErrorCode.INVALID_SURCHARGE_PARAMETER -> TerminalExceptionCodeApi.INVALID_SURCHARGE_PARAMETER
        TerminalErrorCode.READER_COMMUNICATION_SSL_ERROR -> TerminalExceptionCodeApi.READER_COMMUNICATION_SSL_ERROR
        TerminalErrorCode.TAP_TO_PAY_INSECURE_ENVIRONMENT -> TerminalExceptionCodeApi.TAP_TO_PAY_INSECURE_ENVIRONMENT
        TerminalErrorCode.GENERIC_READER_ERROR -> TerminalExceptionCodeApi.UNEXPECTED_READER_ERROR
        TerminalErrorCode.ALLOW_REDISPLAY_INVALID -> TerminalExceptionCodeApi.ALLOW_REDISPLAY_INVALID
        TerminalErrorCode.CANCELED_DUE_TO_INTEGRATION_ERROR -> TerminalExceptionCodeApi.CANCELED_DUE_TO_INTEGRATION_ERROR
        TerminalErrorCode.READER_TAMPERED -> TerminalExceptionCodeApi.READER_TAMPERED
        TerminalErrorCode.PRINTER_BUSY -> TerminalExceptionCodeApi.PRINTER_BUSY
        TerminalErrorCode.PRINTER_PAPERJAM -> TerminalExceptionCodeApi.PRINTER_PAPERJAM
        TerminalErrorCode.PRINTER_OUT_OF_PAPER -> TerminalExceptionCodeApi.PRINTER_OUT_OF_PAPER
        TerminalErrorCode.PRINTER_COVER_OPEN -> TerminalExceptionCodeApi.PRINTER_COVER_OPEN
        TerminalErrorCode.PRINTER_ABSENT -> TerminalExceptionCodeApi.PRINTER_ABSENT
        TerminalErrorCode.PRINTER_UNAVAILABLE -> TerminalExceptionCodeApi.PRINTER_UNAVAILABLE
        TerminalErrorCode.PRINTER_ERROR -> TerminalExceptionCodeApi.PRINTER_ERROR
        TerminalErrorCode.TAP_TO_PAY_UNSUPPORTED_PROCESSOR -> TerminalExceptionCodeApi.TAP_TO_PAY_UNSUPPORTED_PROCESSOR
    }
}

File: stripe_terminal/ios/Classes/Mappings/ExceptionMappings.swift
import Foundation
import StripeTerminal

extension TerminalExceptionApi {
    func toPlatformError() -> PlatformError {
        return PlatformError("mek_stripe_terminal", nil, serialize())
    }
}

extension NSError {
    func toPlatformError(apiError: Error? = nil, paymentIntent: PaymentIntent? = nil) -> PlatformError {
        if (self.domain != "com.stripe-terminal") {
            return PlatformError("\(self.domain):\(self.code)", self.localizedDescription, "\(self)")
        }
        let apiException = toApi(apiError: apiError, paymentIntent: paymentIntent)
        return apiException.toPlatformError()
    }
    
    func toApi(apiError: Error? = nil, paymentIntent: PaymentIntent? = nil) -> TerminalExceptionApi {
        let code = self.toApiCode();
        return TerminalExceptionApi(
            apiError: apiError?.localizedDescription,
            code: code ?? TerminalExceptionCodeApi.unknown,
            message: localizedDescription,
            paymentIntent: paymentIntent?.toApi(),
            stackTrace: nil
        )
    }
    
    private func toApiCode() -> TerminalExceptionCodeApi? {
        let error = ErrorCode(_nsError: self)
        
        switch error.code {
        case .cancelFailedAlreadyCompleted:
            // Ignore this error, the plugin does not allow you to undo an operation more than once
            return nil
        case .notConnectedToReader:
            return .notConnectedToReader
        case .alreadyConnectedToReader:
            return .alreadyConnectedToReader
        case .connectionTokenProviderCompletedWithNothing:
            return nil
        case .connectionTokenProviderCompletedWithNothingWhileForwarding:
            return .connectionTokenProviderErrorWhileForwarding
        case .confirmInvalidPaymentIntent:
            return .confirmInvalidPaymentIntent
        case .nilPaymentIntent:
            return nil
        case .nilSetupIntent:
            return nil
        case .nilRefundPaymentMethod:
            return nil
        case .invalidRefundParameters:
            return .invalidParameter
        case .invalidClientSecret:
            return .invalidClientSecret
        case .invalidDiscoveryConfiguration:
            return nil
        case .invalidReaderForUpdate:
            return .invalidReaderForUpdate
        case .unsupportedSDK:
            return .unsupportedSdk
        case .featureNotAvailableWithConnectedReader:
            return .featureNotAvailableWithConnectedReader
        case .featureNotAvailable:
            return .featureNotEnabledOnAccount
        case .invalidListLocationsLimitParameter:
            return .invalidParameter
        case .bluetoothConnectionInvalidLocationIdParameter:
            return .invalidParameter
        case .invalidRequiredParameter:
            return .invalidRequiredParameter
        case .invalidRequiredParameterOnBehalfOf:
            return .invalidParameter
        case .accountIdMismatchWhileForwarding:
            return .accountIdMismatchWhileForwarding
        case .updatePaymentIntentUnavailableWhileOffline:
            return .updatePaymentIntentUnavailableWhileOffline
        case .updatePaymentIntentUnavailableWhileOfflineModeEnabled:
            return .updatePaymentIntentUnavailableWhileOfflineModeEnabled
        case .forwardingTestModePaymentInLiveMode:
            return .forwardingTestModePaymentInLiveMode
        case .forwardingLiveModePaymentInTestMode:
            return .forwardingLiveModePaymentInTestMode
        case .readerConnectionConfigurationInvalid:
            return .invalidParameter
        case .requestDynamicCurrencyConversionRequiresUpdatePaymentIntent:
            return .requestDynamicCurrencyConversionRequiresUpdatePaymentIntent
        case .dynamicCurrencyConversionNotAvailable:
            return .dynamicCurrencyConversionNotAvailable
        case .readerTippingParameterInvalid:
            return .invalidTipParameter
        case .invalidLocationIdParameter:
            return .invalidParameter
        case .collectInputsInvalidParameter:
            return .invalidParameter
        case .collectInputsUnsupported:
            return .collectInputsUnsupported
        case .canceled:
            return .canceled
        case .locationServicesDisabled:
            return .locationServicesDisabled
        case .bluetoothDisabled:
            return .bluetoothDisabled
        case .bluetoothAccessDenied:
            return .bluetoothPermissionDenied
        case .bluetoothScanTimedOut:
            return .bluetoothScanTimedOut
        case .bluetoothLowEnergyUnsupported:
            return .bluetoothLowEnergyUnsupported
        case .readerSoftwareUpdateFailedBatteryLow:
            return .readerSoftwareUpdateFailedBatteryLow
        case .readerSoftwareUpdateFailedInterrupted:
            return .readerSoftwareUpdateFailedInterrupted
        case .readerSoftwareUpdateFailedExpiredUpdate:
            return .readerSoftwareUpdateFailedExpiredUpdate
        case .bluetoothConnectionFailedBatteryCriticallyLow:
            return .readerBatteryCriticallyLow
        case .cardInsertNotRead:
            return .cardInsertNotRead
        case .cardSwipeNotRead:
            return .cardSwipeNotRead
        case .cardReadTimedOut:
            return .cardReadTimedOut
        case .cardRemoved:
            return .cardRemoved
        case .cardLeftInReader:
            return .cardLeftInReader
        case .offlinePaymentsDatabaseTooLarge:
            return .offlinePaymentsDatabaseTooLarge
        case .readerConnectionNotAvailableOffline:
            return .readerConnectionNotAvailableOffline
        case .readerConnectionOfflineLocationMismatch:
            return .readerConnectionOfflineLocationMismatch
        case .readerConnectionOfflineNeedsUpdate:
            return .readerConnectionOfflineNeedsUpdate
        case .readerConnectionOfflinePairingUnseenDisabled:
            return .locationConnectionNotAvailableOffline
        case .noLastSeenAccount:
            return .noLastSeenAccount
        case .amountExceedsMaxOfflineAmount:
            return .amountExceedsMaxOfflineAmount
        case .invalidOfflineCurrency:
            return .invalidOfflineCurrency
        case .missingEMVData:
            return .missingEmvData
        case .commandNotAllowed:
            return .commandNotAllowed
        case .unsupportedMobileDeviceConfiguration:
            return .unsupportedMobileDeviceConfiguration
        case .passcodeNotEnabled:
            return .passcodeNotEnabled
        case .commandNotAllowedDuringCall:
            return .commandNotAllowedDuringCall
        case .invalidAmount:
            return .invalidAmount
        case .invalidCurrency:
            return .invalidCurrency
        case .appleBuiltInReaderTOSAcceptanceRequiresiCloudSignIn:
            return .appleBuiltInReaderTOSAcceptanceRequiresiCloudSignIn
        case .appleBuiltInReaderTOSAcceptanceCanceled:
            return .appleBuiltInReaderTOSAcceptanceCanceled
        case .collectInputsTimedOut:
            return .collectInputsTimedOut
        case .readerBusy:
            return .readerBusy
        case .incompatibleReader:
            return .incompatibleReader
        case .readerCommunicationError:
            return .readerCommunicationError
        case .nfcDisabled:
            return .nfcDisabled
        case .bluetoothError:
            return .bluetoothError
        case .bluetoothConnectTimedOut:
            return .bluetoothConnectTimedOut
        case .bluetoothDisconnected:
            return .bluetoothDisconnected
        case .bluetoothPeerRemovedPairingInformation:
            return .bluetoothPeerRemovedPairingInformation
        case .bluetoothAlreadyPairedWithAnotherDevice:
            return .bluetoothAlreadyPairedWithAnotherDevice
        case .readerSoftwareUpdateFailed:
            return .readerSoftwareUpdateFailed
        case .readerSoftwareUpdateFailedReaderError:
            return .readerSoftwareUpdateFailedReaderError
        case .readerSoftwareUpdateFailedServerError:
            return .readerSoftwareUpdateFailedServerError
        case .unsupportedReaderVersion:
            return .unsupportedReaderVersion
        case .unknownReaderIpAddress:
            return .unknownReaderIpAddress
        case .internetConnectTimeOut:
            return .internetConnectTimeOut
        case .connectFailedReaderIsInUse:
            return .connectFailedReaderIsInUse
        case .bluetoothReconnectStarted:
            return .bluetoothReconnectStarted
        case .readerNotAccessibleInBackground:
            return .readerNotAccessibleInBackground
        case .appleBuiltInReaderFailedToPrepare:
            return .appleBuiltInReaderFailedToPrepare
        case .appleBuiltInReaderDeviceBanned:
            return .appleBuiltInReaderDeviceBanned
        case .appleBuiltInReaderTOSNotYetAccepted:
            return .appleBuiltInReaderTOSNotYetAccepted
        case .appleBuiltInReaderTOSAcceptanceFailed:
            return .appleBuiltInReaderTOSAcceptanceFailed
        case .appleBuiltInReaderMerchantBlocked:
            return .appleBuiltInReaderMerchantBlocked
        case .appleBuiltInReaderInvalidMerchant:
            return .appleBuiltInReaderInvalidMerchant
        case .appleBuiltInReaderAccountDeactivated:
            return .appleBuiltInReaderAccountDeactivated
        case .readerMissingEncryptionKeys:
            return .readerMissingEncryptionKeys
        case .unexpectedSdkError:
            return .unexpectedSdkError
        case .unexpectedReaderError:
            return .unexpectedReaderError
        case .encryptionKeyFailure:
            return .encryptionKeyFailure
        case .encryptionKeyStillInitializing:
            return .encryptionKeyStillInitializing
        case .collectInputsApplicationError:
            return .collectInputsApplicationError
        case .declinedByStripeAPI:
            return .declinedByStripeApi
        case .declinedByReader:
            return .declinedByReader
        case .commandRequiresCardholderConsent:
            return .customerConsentRequired
        case .refundFailed:
            return .refundFailed
        case .cardSwipeNotAvailable:
            return .cardSwipeNotAvailable
        case .interacNotSupportedOffline:
            return .interacNotSupportedOffline
        case .offlineAndCardExpired:
            return .offlineAndCardExpired
        case .offlineTransactionDeclined:
            return .offlineTransactionDeclined
        case .offlineCollectAndConfirmMismatch:
            return .offlineCollectAndConfirmMismatch
        case .onlinePinNotSupportedOffline:
            return .onlinePinNotSupportedOffline
        case .offlineTestCardInLivemode:
            return .testCardInLiveMode
        case .notConnectedToInternet:
            return .notConnectedToInternet
        case .requestTimedOut:
            return .requestTimedOut
        case .stripeAPIError:
            return .stripeApiError
        case .stripeAPIResponseDecodingError:
            return .stripeApiResponseDecodingError
        case .internalNetworkError:
            return .internalNetworkError
        case .connectionTokenProviderCompletedWithError:
            return .connectionTokenProviderError
        case .connectionTokenProviderCompletedWithErrorWhileForwarding:
            return .connectionTokenProviderErrorWhileForwarding
        case .connectionTokenProviderTimedOut:
            return .connectionTokenProviderTimedOut
        case .sessionExpired:
            return .sessionExpired
        case .notConnectedToInternetAndOfflineBehaviorRequireOnline:
            return .notConnectedToInternetAndOfflineBehaviorRequireOnline
        case .offlineBehaviorForceOfflineWithFeatureDisabled:
            return .offlineBehaviorForceOfflineWithFeatureDisabled
        @unknown default:
            fatalError()
        }
    }
}
